Improved 2017 HCAL Trigger Primitives

matz-e opened the pull request:
Follow up to #15467. With this, the TP implementation should agree with the µHTR spec and clarifications from @jmmans.
### LUT
- Defined magic numbers (mainly bitmasks) in the compression LUTs, compress 2017 TPs properly
- Resize compression LUTs once, not at every DetId
- Add two MSB for fine-grain usage to the linearization LUTs
- Fix linearizations LUTs
### Geometry
- Map TP tower at ieta 26, depth 7 to tower 27
### TP Algo
- Add 2016 HF FG fixes from 80X
- Add a stubb algorithm for 2017 HE FG
- Add proper saturation behavior to HF (according to µHTR spec)
- Proper accounting of overlap tower at ieta 16
- Split 2017 HE TP algorithm from 2009 HE TP algo to handle fine-grain

matz-e committed:
Apply separate threshold for 2016 HCAL FG bit.

Set the correct bit in th TP, too.

Conflicts:
	SimCalorimetry/HcalTrigPrimAlgos/interface/HcalTriggerPrimitiveAlgo.h
	SimCalorimetry/HcalTrigPrimAlgos/src/HcalTriggerPrimitiveAlgo.cc
	SimCalorimetry/HcalTrigPrimProducers/src/HcalTrigPrimDigiProducer.cc

matz-e committed:
Fill in skeleton for QIE11 FG bits and algorithm.

matz-e committed:
Rename Phase1 to 2017.

matz-e committed:
Change fine-grain output signature to a bitset.

matz-e committed:
Add a first algorithm implementation.

matz-e committed:
Remove hard-coded max ADC counts.

matz-e committed:
Fix compilation.

matz-e committed:
Don't consider tower 29 for fine-grain.

matz-e committed:
Saturate towers when one valid channel is saturated.

Also changes LSB by a factor of 4 by changing the energy summation in HF.

matz-e committed:
Explicitly zero HF TP when no valid channels are present.

matz-e committed:
Remove debug statement.

matz-e committed:
Handle tower at ieta 26 properly.

matz-e committed:
Treat towers 16 and 29 right (no FG) and restrict FG depth.

matz-e committed:
Adjust energy handling in the TP algo.

matz-e committed:
Expose some of the HcaluLUTTPGCoder constants.

matz-e committed:
Rework compression for new linearization scale.

matz-e committed:
Fix the range of the upgrade compression LUTs.

The compression function for upgrade HBHE has the same shape as before, but
uses different parameters to achieve the same output size.

matz-e committed:
Fix variable name.

matz-e committed:
Increment bitmask properly.

cmsbuild commented:
A new Pull Request was created by  @matz-e (Matthias Wolf) for CMSSW_8_1_X.

It involves the following packages:

CalibCalorimetry/CaloTPG
CalibCalorimetry/HcalTPGAlgos
CalibFormats/CaloTPG
Geometry/HcalTowerAlgo
SimCalorimetry/HcalTrigPrimAlgos
SimCalorimetry/HcalTrigPrimProducers


@ghellwig, @civanch, @Dr15Jones, @ianna, @mdhildreth, @cmsbuild, @rekovic, @franzoni, @cerminar, @mmusich, @mulhearn, @davidlange6 can you please review it and eventually sign? Thanks.
@ghellwig, @tocheng this is something you requested to watch as well.
@slava77, @smuzaffar you are the release manager for this.

cms-bot commands are list here cms-sw/cmssw#13028

mmusich commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: ecabd61979193d92c659d85df224954508ebee83
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-15735/14956/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-15735/14956/summary.html

The workflows 4.22 have different files in step1_dasquery.log than the ones found in the baseline. You may want to check and retrigger the tests if necessary. You can check it in the "files" directory in the results of the comparisons

matz-e committed:
Merge branch 'CMSSW_8_1_X' into hcal-tp-2017-v8

cmsbuild commented:
Pull request #15735 was updated. @ghellwig, @civanch, @Dr15Jones, @ianna, @mdhildreth, @cmsbuild, @rekovic, @franzoni, @cerminar, @mmusich, @mulhearn, @davidlange6 can you please check and sign again.

matz-e committed:
Adjust trigger modes for recent IBs.

cmsbuild commented:
Pull request #15735 was updated. @ghellwig, @civanch, @Dr15Jones, @ianna, @mdhildreth, @cmsbuild, @rekovic, @franzoni, @cerminar, @mmusich, @mulhearn, @davidlange6 can you please check and sign again.

matz-e commented:
Sorry, had to update to fix compilation on today's IB (after the inclusion of #15679)

mmusich commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

ianna commented on Geometry/HcalTowerAlgo/plugins/HcalTrigTowerGeometryESProducer.cc:
```diff
@@ -23,6 +23,10 @@ HcalTrigTowerGeometryESProducer::produce( const CaloGeometryRecord & iRecord )
+    // TODO this piggy-backs on the HF versioning, which corresponds to
+    // 2017 TP --- trigger modes should really be numbered by year.
+    bool enable2017HE = enable1x1HF;
+    m_hcalTrigTowerGeom->setupHETowers(enable2017HE);
```
@matz-e - this public API is kind of dangerous. An object variable is set based on a topology mode in an object constructed from the topology. The topology will not change, however the API allows it to change effectively allowing perceptive change of the topology. Other objects will see different topology depending on if they retrieve it from the geometry or from the geometry's topology.

matz-e replied:
@ianna, not sure I follow completely. Would it be better to move all this into the constructor of the `HcalTrigTowerGeometry`? I could then remove the `setup{HF,HE}Towers` API calls, and the geometry would be static.

matz-e replied:
@ianna: I have [this commit](https://github.com/matz-e/cmssw/commit/495f70234934de5e5d1527355714ab55d79d2b1c) lined up to address this in a follow-up PR.

cmsbuild commented:
+1
Tested at: 3e0cf292bab4c30a20edc5c492c753dd9f1d0586
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-15735/14980/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/8c9f703966a32f2110220e24fe140865a8ee57ee
https://github.com/cms-sw/cmssw/commit/0e99a5c21be9ebaba2f33c838fe7908d15091b4d
https://github.com/cms-sw/cmssw/commit/19be1746f255207b65aa74c659a522e2acf69b0f
https://github.com/cms-sw/cmssw/commit/8d4002f769859ac02e89c44074183add883f104d
https://github.com/cms-sw/cmssw/commit/7ad1bce3f4a3011ac8543c79847d7be92cc2bb80
https://github.com/cms-sw/cmssw/commit/6d5e42c79913759d564bcbe787a141b85a3d8a51
https://github.com/cms-sw/cmssw/commit/ed88e5a443d68f4bff6a4a84b0a0c8e267095bd9
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-15735/14980/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-15735/14980/git-merge-result

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-15735/14980/summary.html

ianna commented:
+1

matz-e committed:
Seal HcalTrigTowerGeometry.

Don't allow the trigger tower geometry settings to diverge from the
topology via the API.

abdoulline commented:
Dear colleagues, we need it in pre12 ...
 @ghellwig,  @rekovic, @mmusich, @mulhearn

mmusich commented:
@abdoulline I think for the changes in `CalibCalorimetry/CaloTPG` we need an explicit sign-off from our L1T colleagues before moving on.

matz-e commented:
That would be @mulhearn or @rekovic, right?

abdoulline commented:
@mulhearn  @rekovic  
we just cannot stay idle so long in preparation of major 2017 HCAL changes

davidlange6 commented on CalibCalorimetry/CaloTPG/src/CaloTPGTranscoderULUT.cc:
```diff
@@ -66,32 +62,30 @@ void CaloTPGTranscoderULUT::loadHCALCompress(HcalLutMetadata const& lutMetadata,
+	for (unsigned int i = 0; i < threshold; ++i) outputLUT_[index][i] = 0;
+	for (unsigned int i = threshold; i < lutsize; ++i){
+           LUT value;
+           if (isHBHE and lutsize == QIE8_OUTPUT_LUT_SIZE) {
```
@matz-e - move this set of if statements outside of the look over lutsize

Second - it would be more clear if analytical*LUT variables were renamed to reflect their physical meaning (eg, analyticalLUTQIE8 ?)

matz-e committed:
Rename analytical LUTs to reflect their content.

matz-e committed:
Switch order of conditions and loops.

cmsbuild commented:
Pull request #15735 was updated. @ghellwig, @civanch, @Dr15Jones, @ianna, @mdhildreth, @cmsbuild, @rekovic, @franzoni, @cerminar, @mmusich, @mulhearn, @davidlange6 can you please check and sign again.

matz-e commented:
@davidlange6 does this address your comments? I've also included @ianna's requested API change.

davidlange6 commented:
please test
(looks ok to me)

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: d4b763b95be8ae7ed9d64babaa35ebeb1198f9c4
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-15735/15153/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-15735/15153/summary.html

ianna commented:
+1

abdoulline commented:
@rekovic, @franzoni, @mmusich, @mulhearn
may we have your attention, please?

rekovic commented:
+1

mmusich commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_8_1_X IBs (tests are also fine). This pull request requires discussion in the ORP meeting before it's merged. @slava77, @davidlange6, @smuzaffar

davidlange6 commented:
+1
